light: update initialization description (#320)
tendermint/spec: spec/light-client/supervisor/supervisor_001_draft.md

detection of an attack yet instead requires an initial point of trust.
There are three forms of initial data:

- A trusted block from a prior initialization
- A trusted height and hash
- A genesis file

The golang light client implementation checks this initial data in that 
order; first attempting to find a trusted block from the trusted store, 
then acquiring a light block from the primart at the trusted height and matching
the hash, or finally checking for a genesis file to verify the initial header.

The light client doesn't need to check if the trusted block is within the
trusted period because it already trusts it, however, if the light block is 
outside the trust period, there is a higher chance the light client won't be
able to verify anything. 

Cross-checking this trusted block with providers upon initialization is helpful
for ensuring liveness and correct configuration of such nodes but does not 
increase trust since proving a conflicting block is a 
[light client attack](https://informal.systems) 
and not just a [bogus](https://informal.systems) block could result in
performing backwards verification beyond the trusted period, thus a fruitless

> endeavour.

Author: Here I do note that cross-checking does no increase trust. I then go on to say that the `golang light client imlementation` performs a "consistency check" on all providers

Reviewer: Sorry, I didn't see that earlier.

I am not sure I understand how it is helpful for liveness, though? Perhaps you could add a sentence that clarifies the scenario you have in mind?

Author: By liveness I'm meaning that we were able to establish a connection with the provider and they were responsive. This might actually be a different definition of liveness to the one you have in mind

Reviewer: I didn't have any definition in mind ;-)  
I think if there is a specific goal, it should be written more explicitly.

    switch {
    case LCInitData.TrustedBlock != nil:
        // we trust the block from a prior initialization
        initialBlock = LCInitData.TrustedBlock

    case LCInitData.TrustedHash != nil:
        untrustedBlock := FetchLightBlock(PeerList.Primary(), LCInitData.TrustedHeight)
        

        // verify that the hashes match
        if untrustedBlock.Hash() != LCInitData.TrustedHash {
            return nil, Error("Primary returned block with different hash")
        }
        // after checking the hash we now trust the block
        initialBlock = untrustedBlock        
    }
    case LCInitData.Genesis != nil:
        untrustedBlock := FetchLightBlock(PeerList.Primary(), LCInitData.Genesis.InitialHeight)
        
        // verify that 2/3+ of the validator set signed the untrustedBlock
        if err := VerifyCommit(untrustedBlock.Commit, LCInitData.Genesis.Validators); err != nil {
            return nil, err
        }

        // we can now trust the block
        initialBlock = untrustedBlock
    default:
        return nil, Error("No initial data was provided")

    if err := CrossCheck(initialBlock, PeerList.Witnesses()); err != nil {
        return nil, err

>     }

Author: I could remove this cross check from the spec pseudocode as it's not really important to the correctness of the spec. I think `[LC-FUNC-INIT.1]` would need to be updated to  `[LC-FUNC-INIT.2]` anyway because there are still changes to the logic

Reviewer: OK. Perhaps we could also add a comment within the pseudo code that that line is optional and done in the golang version. I am a bit under water right now and will do a closer check of the new version in the week of July 26

the hash, or finally checking for a genesis file to verify the initial header.

The light client doesn't need to check if the trusted block is within the
trusted period because it already trusts it, however, if the light block is 
outside the trust period, there is a higher chance the light client won't be
able to verify anything. 

Cross-checking this trusted block with providers upon initialization is helpful
for ensuring liveness and correct configuration of such nodes but does not 
increase trust since proving a conflicting block is a 
[light client attack](https://informal.systems) 
and not just a [bogus](https://informal.systems) block could result in
performing backwards verification beyond the trusted period, thus a fruitless
endeavour.

However, with the notion of it's better to fail earlier than later, the golang
light client implementation will perform a consistency check on all providers
and will error if one returns a different header, allowing the user
the opportunity to reinitialize.

#### **[LC-FUNC-INIT.1]:**

```go
func InitLightClient(initData LCInitData) (LightStore, Error) {
    var initialBlock LightBlock

    switch {
    case LCInitData.TrustedBlock != nil:

>         // we trust the block from a prior initialization

Reviewer: Is it necessarily from "prior" initialization?

Author: I guess not. As in it's feasible that someone adds a light block to the store and then passes that in (without it going through a previous initialization process)

Reviewer: I just wondered, because the text of the comment was different in the previous version of the spec?

Author: Right. The way the light client is initialised, at least in the golang repo, is that we give users the option to either start form a trusted hash and height or from a trusted store which would've come from previously using the light client. I guess I was just translating that notion across

Reviewer: It would be great to add this explanation as a comment to the spec.

        initialBlock = LCInitData.TrustedBlock

>     case LCInitData.TrustedHash != nil:

Reviewer: Please add the Field `TrustedHash` to the data structure definition further above in the code.

Author: Sure. I've added them. One might argue that the `TrustedHash` and `TrustedHeight` combination are just for convenience (instead of passing the entire `TrustedBlock`) and so doesn't belong in the spec. But perhaps it's good to have them here just for documentation sake because it's definitely a prevalent concept in the golang codebase

Reviewer: I think it is good to give the options and details here. Thanks for adding the hash option?

        untrustedBlock := FetchLightBlock(PeerList.Primary(), LCInitData.TrustedHeight)
        

        // verify that the hashes match
        if untrustedBlock.Hash() != LCInitData.TrustedHash {
            return nil, Error("Primary returned block with different hash")
        }
        // after checking the hash we now trust the block
        initialBlock = untrustedBlock        
    }
    case LCInitData.Genesis != nil:
        untrustedBlock := FetchLightBlock(PeerList.Primary(), LCInitData.Genesis.InitialHeight)
        
        // verify that 2/3+ of the validator set signed the untrustedBlock

>         if err := VerifyCommit(untrustedBlock.Commit, LCInitData.Genesis.Validators); err != nil {

Reviewer: Is the function defined somewhere? Previously we used
[`ValidandVerified`](https://github.com/tendermint/spec/blob/master/spec/light-client/verification/verification_002_draft.md#lcv-func-valid2) here (with a typo). As it takes as inputs two lightblocks we generated a gensisblock.

Author: It's mentioned here: https://github.com/tendermint/spec/blob/master/spec/light-client/verification/README.md and is the name of a function within the golang codebase. 

AFAICS: `ValidandVerified` does three things
- Check that all formats are consistent and as expected
- Checks that their is a sufficient degree of overlap in the validators between two heights
- Checks that the 2/3+ of the validator set and the untrusted height signed the untrusted block

`VerifyCommit` performs only the last of these three things

Reviewer: I see. I think we should add the function definition of `VerifyCommit` to the spec. (Also the linked README.md only contains `VerifyCommitTrusting` and `VerifyCommitFull`)

Author: In this case, I'm talking about `VerifyCommitFull`. `VerifyCommitTrusting` is the name of the function that performs the second bullet point

Author: Should I add these functions to the verification document?

Reviewer: Yes, please

Author: Added `VerifyCommitLight` here: https://github.com/tendermint/spec/blob/callum/light-client/spec/light-client/verification/verification_001_published.md#lcv-func-verifycommitlight1

Author: It would be nice to have all these verification function well documented in the spec: https://github.com/tendermint/tendermint/blob/master/types/validation.go

#### Initialization

The light client is based on subjective initialization. It has to
trust the initial data given to it by the user. It cannot perform any
detection of an attack yet instead requires an initial point of trust.
There are three forms of initial data which are used to obtain the 
first trusted block:

- A trusted block from a prior initialization

> - A trusted height and hash

Reviewer 2: Is height necessary? Wouldn't `hash` identify a block uniquely?

Reviewer: I guess the problem is that currently we can only ask a full node for a block at a given height, rather than a block corresponding to a hash.

Reviewer 2: I see, thank you.

- A genesis file

The golang light client implementation checks this initial data in that
order; first attempting to find a trusted block from the trusted store,
then acquiring a light block from the primary at the trusted height and matching
the hash, or finally checking for a genesis file to verify the initial header.

The light client doesn't need to check if the trusted block is within the
trusted period because it already trusts it, however, if the light block is
outside the trust period, there is a higher chance the light client won't be

> able to verify anything.

Reviewer 2: If the trusted block is outside of the "trust period" (i.e. all of its validators have already unbonded), can it be trusted at all? Light client would be subject to a long range attack in that case, correct?

Cross-checking this trusted block with providers upon initialization is helpful

> for ensuring that the node is responsive and correctly congfigured but does not

Reviewer 2: ```suggestion
for ensuring that the node is responsive and correctly configured but does not
```

increase trust since proving a conflicting block is a

> [light client attack](https://informal.systems)

Reviewer: ```suggestion
[light client attack](https://github.com/tendermint/spec/blob/master/spec/light-client/detection/detection_003_reviewed.md#tmbc-lc-attack1)
```

> and not just a [bogus](https://informal.systems) block could result in

Reviewer 2: I'm not sure how these links are helping here.

Reviewer: Thanks a lot for spotting the default links!! I have replaced them with meaningful ones.

Reviewer 2: Thank you!

> and not just a [bogus](https://informal.systems) block could result in

Reviewer: ```suggestion
and not just a [bogus](https://github.com/tendermint/spec/blob/master/spec/light-client/detection/detection_003_reviewed.md#tmbc-bogus1) block could result in
```
